src/openms_gui/source/VISUAL/TVToolDiscovery.cpp

```diff
@@ -55,87 +57,182 @@ namespace OpenMS
-    auto return_state = ExternalProcess().run(executable.toQString(), args, working_dir.toQString(), false, ExternalProcess::IO_MODE::NO_IO);
-    --running_processes;
+    auto lam_out = [&](const String& out) { OPENMS_LOG_INFO << out; };
+    auto lam_err = [&](const String& out) { OPENMS_LOG_INFO << out; };
+    ExternalProcess proc(lam_out, lam_err);
+    auto return_state = proc.run(executable.toQString(), args, working_dir.toQString(), true, ExternalProcess::IO_MODE::NO_IO);
     // Return empty param if writing the ini file failed
     if (return_state != ExternalProcess::RETURNSTATE::SUCCESS)
     {
+      OPENMS_LOG_WARN << "TOPP tool: " << executable << " error during execution: " << (uint32_t)return_state << "\n";
       return tool_param;
     }
     // Parse ini file to param object
     ParamXMLFile paramFile;
-    paramFile.load((path).c_str(), tool_param);
+    try
+    {
+      paramFile.load((path).c_str(), tool_param);
+    }
+    catch(const Exception::FileNotFound& e)
+    {
+      OPENMS_LOG_WARN << e << "\n" << "TOPP tool: " << executable <<
+        " not able to write ini. Plugins must implement -write_ini parameter. Skipping." << std::endl;
+      return tool_param;
+    }
+    
+    if (plugins) {
+      auto tool_name = tool_param.begin().getTrace().begin()->name;
+      auto filename = tool_path.substr(tool_path.find_first_of('/') + 1);
+      tool_param.setValue(tool_name + ":filename", filename, "The filename of the plugin executable, this entry is automatically generated");
+    }
+
     return tool_param;
   }
+
+  // MAYBE USE THIS TO GET THE PLUGINS IN TOOLSDIALOG
+  const std::vector<std::string>& TVToolDiscovery::getPlugins()
+  {
+    return plugins_;
+  }
+
+  const StringList TVToolDiscovery::getPlugins_()
+  {
+    StringList plugins;
+
+    // here all supported file extensions can be added
+    std::vector<std::string> valid_extensions {"", ".py"};
+    const auto comparator = [valid_extensions](const std::string& plugin) -> bool
+    {
+        return !File::executable(plugin) ||
+          (std::find(valid_extensions.begin(), valid_extensions.end(), std::filesystem::path(plugin).extension()) == valid_extensions.end());
+    };
+
+    if (File::fileList(plugin_path_, "*", plugins, true))
+    {
+      plugins.erase(std::remove_if(plugins.begin(), plugins.end(), comparator), plugins.end());
+    }
+
+    return plugins;
+  }
+
+  void TVToolDiscovery::setPluginPath(const std::string &path)
+  {
+    plugin_path_ = path;
+  }
+
+  std::string TVToolDiscovery::findPluginExecutable(const std::string &name)
+  {
+    //TODO: At the moment the usage of subdirectories in the plugin path are not possible
+    //To change that, the tool scanner has to recursively search all directories in the plugin path
+    if (!plugin_params_.exists(name + ":filename"))
+    {
+      return "";
+    }
+    return plugin_path_ + "/" + plugin_params_.getValue(name + ":filename").toString();
```

Comment:
safe on windows?

Reply:
I can't test on windows so I don't know.

Reply:
should be ok